### PR TITLE
Add UI for modifying chunk options

### PR DESCRIPTION
**Pull request by kevinushey**
Not quite ready for merge; needs discussion to refine UI + available options.

This PR adds a 'gear' icon that allows users to modify chunk options more easily:

chunk-options

The idea is that changes in the popup immediately synchronize with the displayed text.

## TODO
- [x] Fix focus issues when clicking a gear when source editor not in focus
- [x] Consider other, non-boolean options (figure width / height? engine?)
- [x] Consider three-state checkbox to allow for [empty | off | on] states
- [x] Styling?
- [x] Display over text

### CODE REVIEW

**Comment by kevinushey** [comment]
Here's how it looks now, with placeholder styling for the tri-state checkbox:

screen shot 2015-05-21 at 3 31 08 pm

I also added a tiny bit of styling to the icons (just so they have their own background where they're displayed)

After I fix up the focus issue mentioned earlier + get some new images / styling for the icons I think it should be good for merge.

**Comment by jjallaire** [comment]
This is looking very good!

I'd change the text to (removing all of the question marks as they are
implied):

Evaluate R code

Print R output into document

Include R code in document

**Comment by kevinushey** [comment]
Almost ready to go; just need to fix one more bug re: display here:

screen shot 2015-05-21 at 4 13 58 pm

I think I'll just make the background opaque for the icons and pick an appropriate color depending on the theme.

**Comment by jjallaire** [comment]
Just sent Paul a note regarding our outstanding UI questions so hopefully we'll have that resolved and incorporated today. Will merge after we get those tweaks in. Note that I still recommend that we tweak the language. Here's a variation that uses the actual chunk option as the first word of each sentence:

```
Evaluate R code
Include chunk output in document
Echo R code into document
```

File: src/gwt/src/org/rstudio/studio/client/RStudioGinModule.java [hunk 1]
```diff
@@ -365,6 +366,8 @@ protected void configure()
 
       bind(DocDisplay.class).to(AceEditor.class);
       
+      bind(ChunkIconsManager.class).in(Singleton.class);
```
Comment (jjallaire) [inline src/gwt/src/org/rstudio/studio/client/RStudioGinModule.java]: Not sure that this needs to be a Singleton (is there state we need to share across invocations?).

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkIconsManager.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+import com.google.gwt.user.client.ui.Widget;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+@Singleton
+public class ChunkIconsManager
+{
+   public ChunkIconsManager()
+   {
+      RStudioGinjector.INSTANCE.injectMembers(this);
```
Comment (jjallaire) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkIconsManager.java]: Perhaps it's this event handler that's the "shared" element which pushes this into being a singleton?

**Comment by gokceneraslan** [comment]
How about displaying the gear icon of a chunk only if the mouse is hovering over the area of the chunk in question?

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkIconsManager.java (hunk 2 above)
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkIconsManager.java]: Yeah, that's basically why -- not sure if this is the best way of handling it (e.g. in the future when we have multiple source documents / windows open, should we have one instance per window?) but I think it's reasonable, at least for now.

**Comment by kevinushey** [comment]
I think we want it visible by default for now (just to increase exposure to users) but we should definitely have an option for hiding / intelligently displaying the toolbar.

**Comment by kevinushey** [comment]
Current treatment:

screen shot 2015-05-22 at 12 34 26 pm

@jjallaire, I didn't use the `ThemedButton`s because they were to large to fit comfortably here, and it doesn't look like there's a simple way to resize.

There's also a subtle issue where the background colouring of our themed popup panel doesn't work great -- this is because the border includes an opaque white, and that white overlays the background colour we attempt to set.

Do you think it's worth attempting to fix that? (setting white to transparent; explicitly declaring background colors elsewhere as needed). That could definitely get messy though.

**Comment by jjallaire** [comment]
Are you using SmallButton? I think that will give you the size /
form-factor you want and still be themed consistently across
browsers/platforms.

We should probably talk realtime about the issues surrounding the
background coloring. This might be worth addressing however not if as you
said gets too messy.

I think we should change the text of the options to read as follows (note
the first word of each sentence matches the chunk option):

Evaluate R code
Include chunk output
Echo R code

If you want them to be a bit longer you can use:

Evaluate R code
Include chunk output in document
Echo R code in document

But if the first set looks okay I'd go with that.

**Comment by kevinushey** [comment]
Okay, I think this is good enough for merge (there may be some behaviour refinement still needed but I think it's ready for master + initial testing in the dailies)

I added a simple `MiniPopupPanel` class that acts like the `MiniDialogPopupPanel`, but without the dialog header + the minimal styling we discussed. I also added some 'smart' drag handling so you can click popups any place without an 'input' handler and drag -- @jjallaire, do you want to sanity-check my implementation?

Here's what it looks like now:

screen shot 2015-05-22 at 1 50 56 pm

I also added a display option to toggle display of chunk icons, and added some color utilities for managing the background color of the inline toolbar.

**Comment by kevinushey** [comment]
Redirect for the Chunk Options help link: https://github.com/rstudio/rstudio-web/pull/4

**Pull request merged**